plugins/flow-visibility/clickhouse-monitor/main.go

```diff
@@ -0,0 +1,234 @@
+			podLogs, err := req.Stream(context.TODO())
+			if err != nil {
+				return logString, fmt.Errorf("error in opening stream: %v", err)
+			}
+			defer podLogs.Close()
+
+			buf := new(bytes.Buffer)
+			_, err = io.Copy(buf, podLogs)
+			if err != nil {
+				return logString, fmt.Errorf("error in copy information from podLogs to buf: %v", err)
+			}
+			logString := buf.String()
+			return logString, nil
+		}
+	}
+	return logString, fmt.Errorf("failed to find a succeeded monitor")
+}
+
+// Connects to Clickhouse in a loop
+func connectLoop() (*sql.DB, error) {
+	// Connection to Clickhouse timeout if if fails for 1 minute.
+	connectionTimeout := time.Minute
+	// Retry connection to Clickhouse every 5 seconds if it fails.
+	connectionWait := 5 * time.Second
+
+	// Clickhouse configuration
+	userName := os.Getenv("CH_USERNAME")
+	password := os.Getenv("CH_PASSWORD")
+	svcAddress := "clickhouse-clickhouse.flow-visibility.svc.cluster.local"
+	svcPort := 9000
+
+	ticker := time.NewTicker(connectionWait)
+	defer ticker.Stop()
+
+	timeoutExceeded := time.After(connectionTimeout)
+	for {
+		select {
+		case <-timeoutExceeded:
+			return nil, fmt.Errorf("failed to connect to Clickhouse after %s", connectionTimeout)
+
+		case <-ticker.C:
+			// Open the database and ping it
+			dataSourceName := fmt.Sprintf("tcp://%s:%d?debug=true&username=%s&password=%s", svcAddress, svcPort, userName, password)
+			connect, err := sql.Open("clickhouse", dataSourceName)
+			if err != nil {
+				klog.Errorf("Failed to connect to Clickhouse: %v", err)
+			}
+			if err := connect.Ping(); err != nil {
+				if exception, ok := err.(*clickhouse.Exception); ok {
+					klog.Errorf("Failed to ping Clickhouse: %s", exception.Message)
+				} else {
+					klog.Errorf("Failed to ping Clickhouse: %v", err)
+				}
+			} else {
+				return connect, nil
+			}
+		}
+	}
+}
+
+// Checks the memory usage in the Clickhouse, deletes records when it exceeds the threshold.
+func monitorMemory(connect *sql.DB) bool {
+	// Get memory usage from Clickhouse system table
+	rows, err := connect.Query("SELECT free_space, total_space FROM system.disks")
+	if err != nil {
+		klog.Errorf("Failed to get memory usage for Clickhouse: %v", err)
+		return false
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var (
+			freeSpace  uint64
+			totalSpace uint64
+		)
+		if err := rows.Scan(&freeSpace, &totalSpace); err != nil {
+			klog.Errorf("Failed to read memory usage for Clickhouse: %v", err)
+			return false
+		}
+		// Calculate the memory usage
+		usagePercentage := float64(totalSpace-freeSpace) / float64(totalSpace)
+		klog.Infof("Memory usage: total %d, used: %d, percentage: %f", totalSpace, totalSpace-freeSpace, usagePercentage)
+		// Delete records when memory usage is larger than threshold
+		if usagePercentage > threshold {
+			command := fmt.Sprintf("ALTER TABLE %s DELETE WHERE timeInserted IN (SELECT timeInserted FROM %s LIMIT %d)", tableName, tableName, getDeleteRowNum(connect))
```

Comment:
I think using `WHERE timeInserted IN (a set of timeinserted value)` may be inefficient, could you try optimizing to use `WHERE timeInserted < an upper boundary of timeinserted` instead?

Reply:
Done